File: 3702-maximum-subarray-with-equal-products/3702-maximum-subarray-with-equal-products.py
class Solution:
    def maxLength(self, nums: List[int]) -> int:
        def lcm(a, b):
            return (a * b) // gcd(a, b)

        ans = 0
        for i in range(len(nums)):
            lcm_v = gcd_v = prod_v = nums[i]

            for j in range(i + 1, len(nums)):
                prod_v *= nums[j]
                gcd_v = gcd(gcd_v, nums[j]) #Inbuilt in Python 
                lcm_v = lcm(lcm_v, nums[j]) # Defined Above
                if prod_v == (gcd_v * lcm_v):
                    ans = max(ans, j - i + 1)
        return ans

File: 3702-maximum-subarray-with-equal-products/3702-maximum-subarray-with-equal-products.java
class Solution {
    public int maxLength(int[] nums) {
        int ans = 0;
        for (int i = 0; i < nums.length; i++) {
            long gcdV = nums[i], lcmV = nums[i], prodV = nums[i];

            for (int j = i + 1; j < nums.length; j++) {
                gcdV = gcd(gcdV, nums[j]);
                prodV *= (long) nums[j];
                lcmV = lcm(lcmV, (long) nums[j]);

                if (prodV == gcdV * lcmV) {
                    ans = Math.max(ans, j - i + 1);
                }
            }
        }
        return ans;
    }

    public long gcd(long a, long b) {
        if (b == 0)
            return a;
        return gcd(b, a % b);
    }

    public long lcm(long a, long b) {
        return (a * b) / gcd(a, b);
    }
}